tools/flang1/utils/n2rst/CMakeLists.txt

```diff
@@ -46,7 +48,8 @@ add_custom_command(
                                              ${FLANG1_DOC_SRC_DIR}/semant.n
                                              ${FLANG1_DOC_SRC_DIR}/transform.n
   WORKING_DIRECTORY ${FLANG1_DOC_BIN_DIR}
-  DEPENDS fen2rst ${FLANG1_DOC_SRC_DIR}/ast.n
+  DEPENDS fen2rst ${UTILS_AST_DIR}/ast.n
+                  ${FLANG1_DOC_SRC_DIR}/commopt.n
```

Comment:
My build is failing with this error message:
```
[ 14%] Built target fen2rst
make[2]: *** No rule to make target 'tools/flang1/utils/prstab/gram.lis', needed by 'docs/sphinx/flang1/ast.rst'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:1806: tools/flang1/utils/n2rst/CMakeFiles/flang1_gen_sphinx_docs.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

Reply:
I fixed the above error with this patch:
```diff
diff --git a/tools/flang1/utils/prstab/CMakeLists.txt b/tools/flang1/utils/prstab/CMakeLists.txt
index 59a81011..c9837a51 100644
--- a/tools/flang1/utils/prstab/CMakeLists.txt
+++ b/tools/flang1/utils/prstab/CMakeLists.txt
@@ -21,7 +21,7 @@ add_executable(prodstr
 # Generate parse tables

 add_custom_command(
-  OUTPUT ${UTILS_PARSETABLE_BIN_DIR}/gramdf.h ${UTILS_PARSETABLE_BIN_DIR}/gramsm.h ${UTILS_PARSETABLE_BIN_DIR}/gramtk.h ${UTILS_PARSETABLE_BIN_DIR}/proddf.h ${UTILS_PARSETABLE_BIN_DIR}/tokdf.h ${UTILS_PARSETABLE_BIN_DIR}/gramnt.h
+  OUTPUT ${UTILS_PARSETABLE_BIN_DIR}/gramdf.h ${UTILS_PARSETABLE_BIN_DIR}/gramsm.h ${UTILS_PARSETABLE_BIN_DIR}/gramtk.h ${UTILS_PARSETABLE_BIN_DIR}/proddf.h ${UTILS_PARSETABLE_BIN_DIR}/tokdf.h ${UTILS_PARSETABLE_BIN_DIR}/gramnt.h ${UTILS_PARSETABLE_BIN_DIR}/gram.lis
   # Copy to current build directory, otherwise header will be generated in
   # original directory
   COMMAND ${CMAKE_COMMAND} -E copy ${UTILS_PARSETABLE_DIR}/gram.txt ${UTILS_PARSETABLE_BIN_DIR}
```

Reply:
Thanks for pointing this out @bryanpkc. Indeed, I managed to reproduce the issue when I `rm -rf`-ed the build directory and started from scratch.
Unfortunately your patch did not improve the situation.  
It might matter that I build with `-DFLANG_INCLUDE_DOCS=ON`, which is the flag I found to be used in the CMake files. I admit I am currently looking mainly at HTML docs generated by Sphinx and this flag did the job for me. I see no explicit documentation on how docs should be built - perhaps that would be an idea for another improvement? I doubt that the nroff/rst files influence the doxygen build.
When I used the flags you mentioned I got a different error. But then... maybe I should build `llvm` with these flags as well?

Anyway, regarding `gram.lis` - this file doesn't exist anywhere in the source tree and is not produced during build process.  Also - the sphinx docs will eventually use `flang1/grammar.rst` and `gram.rst` would remain unused. `git blame` shows that [most of this file remained intact](https://github.com/flang-compiler/flang/blame/master/tools/flang1/utils/n2rst/CMakeLists.txt) since the beginning of this repo, so perhaps it was never really built?

I suggest we remove the lines which mention `gram.list` and `gram.rst` from the CMakeLists.txt. Would that be OK, @bryanpkc ?

Reply:
@michalpasztamobica I wasn't discussing Doxygen at all.

I am building master branch, plus the patch above, with `-DLLVM_INCLUDE_DOCS=on` and `make -j4 install`, after deleting the build directory, and I get the correct Sphinx files. There is currently a race condition in the Sphinx command which requires running `make -j4 install` a second time in order to generate all the files and index.html correctly. I haven't needed to build LLVM any differently (I did not add any docs-related option when running `cmake` for llvm-project), but I should add that I am using release_100 classic-flang-llvm-project.

`gram.lis` is still being generated by the `lr` command in my build. You can find it under `$CMAKE_BINARY_DIR/tools/flang1/utils/prstab/`; it contains the actual text for the grammar, and Sphinx generates `gram.html` from it. In contrast, `tools/flang1/docs/grammar.n`contains just a pointer to `gram.html`. Do you have a different `grammar.n` than the master branch?

P.S. I agree we should update the "Building Flang" wiki page with a sentence or two on how to build the docs.

Reply:
It took me a while, but now I can reliably build all the docs in a single run even with `make -j6`.

You were right, @bryanpkc : `gram.html` is being generated and used to produce `grammar.html`. I included your patch. Thanks!

Regarding the two DOCS flags and the wiki update... I am not entirely sure but to my understanding  `LLVM_INCLUDE_DOCS` [is being set if](https://github.com/flang-compiler/flang/blob/a6d7701ea4f4ea37b09ac83d8351623d0292c489/CMakeLists.txt#L417) `FLANG_INCLUDE_DOCS` is set. The latter is also a valid CMake option. I don't have the privileges to edit [wiki with build instructions](https://github.com/flang-compiler/flang/wiki/Building-Flang), but perhaps you or @kiranchandramohan could put up a sentence saying "To build the the HTML and Doxygen documentation add `-DFLANG_INCLUDE_DOCS` to the cmake command" somewhere in point 7 of the wiki page.

Reply:
Thanks @michalpasztamobica!

> LLVM_INCLUDE_DOCS is being set if FLANG_INCLUDE_DOCS is set.

It's the other way around... The default value for `FLANG_INCLUDE_DOCS` is the value of `LLVM_INCLUDE_DOCS`. Setting either will work. Setting the latter would be useful if we were building the Flang project in-tree (but we are not).

Reply:
You're right @bryanpkc , I wrote it the wrong way round, thanks for correction.

Reply:
I have updated the [Building-Flang](https://github.com/flang-compiler/flang/wiki/Building-Flang) page.